### PR TITLE
feat: Add support for OperationContextParams

### PR DESCRIPTION
**Pull request by sichanyoo**
Companion PR
- https://github.com/smithy-lang/smithy-swift/pull/800

## Issue \#

- https://github.com/awslabs/aws-sdk-swift/issues/1517
- Corresponding SIM: `SWIFT-1125`

## Description of changes

- Add support for endpoint parameters that have `@operationContextParams` trait
- Update codegen tests

## Conventional Commits

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: codegen/Package.swift [hunk 1]
```diff
@@ -154,6 +154,7 @@ func addProtocolTests() {
         .init(name: "EventStream", sourcePath: "\(baseDirLocal)/EventStream", buildOnly: true),
         .init(name: "RPCEventStream", sourcePath: "\(baseDirLocal)/RPCEventStream", buildOnly: true),
         .init(name: "Waiters", sourcePath: "\(baseDirLocal)/Waiters", testPath: "../codegen/protocol-test-codegen-local/Tests"),
+        .init(name: "StringArrayEndpointParam", sourcePath: "\(baseDirLocal)/StringArrayEndpointParam")
```
Comment (sichanyoo) [inline codegen/Package.swift:157]: Add new target for running runtime test that run against generated code

File: codegen/protocol-test-codegen-local/build.gradle.kts [hunk 2]
```diff
@@ -47,6 +47,10 @@ val codegenTests = listOf(
     CodegenTest(
         "aws.protocoltests.eventstream#RPCTestService",
         "RPCEventStream"
+    ),
+    CodegenTest(
+        "aws.endpointtests.stringarray#EndpointStringArray",
+        "StringArrayEndpointParam"
```
Comment (sichanyoo) [inline codegen/protocol-test-codegen-local/build.gradle.kts:53]: Add the endpoint string array Smithy model to list of models used by this build script

File: codegen/protocol-test-codegen-local/model/endpoints-string-array.smithy [hunk 3]
```diff
@@ -0,0 +1,197 @@
+$version: "2.0"
```
Comment (sichanyoo) [inline codegen/protocol-test-codegen-local/model/endpoints-string-array.smithy:1]: This is the Smithy model file provided by SEP that has prescribed test cases. Only edits made were putting on `@service` trait and `@restJson1` trait to the service shape, and adding `@http` traits to each of the operation shape with dummy values. This was done to "plug" this into existing protocol test codegen flow. Building the local protocol test gradle project generates Swift service code for this model, along with the endpoint tests.

File: codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/EndpointTestGenerator.kt [hunk 4]
```diff
@@ -66,7 +66,7 @@ class EndpointTestGenerator(
                             val value = Value.fromNode(pair.second)
```
Comment (sichanyoo) [inline codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/EndpointTestGenerator.kt:66]: Changes in this file fixes a bug in endpoint test generator where it was incorrectly & unnecessarily casting array endpoint params to `[AnyHashable]`. This was never caught before bc we never had test for it & no service has string array endpoint params yet.

File: codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/middleware/OperationEndpointResolverMiddleware.kt [hunk 5]
```diff
@@ -6,26 +6,35 @@
 package software.amazon.smithy.aws.swift.codegen.middleware
```
Comment (sichanyoo) [inline codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/middleware/OperationEndpointResolverMiddleware.kt:6]: Changes in this file add support for handling the `@operationContextParams` trait

File: codegen/smithy-aws-swift-codegen/src/test/kotlin/software/amazon/smithy/aws/swift/codegen/EndpointParamsGeneratorTests.kt [hunk 6]
```diff
@@ -30,33 +30,45 @@ public struct EndpointParams {
     public let boolBaz: Swift.String?
```
Comment (sichanyoo) [inline codegen/smithy-aws-swift-codegen/src/test/kotlin/software/amazon/smithy/aws/swift/codegen/EndpointParamsGeneratorTests.kt:30]: From here on it's just codegen test updates

**Review by sichanyoo — COMMENTED** [review]
Comments to help reviewers

**Review by jbelkins — APPROVED** [review]

**Pull request merged**